persistent/cPersistence.c

```diff
@@ -190,47 +192,54 @@ ghostify(cPersistentObject *self)
         *dictptr = NULL;
     }
 
-    /* clear all slots besides _p_* */
-    slotnames = pickle_slotnames(Py_TYPE(self));
-    if (slotnames && slotnames != Py_None)
+    /* clear all slots besides _p_*
+     * ( for backward-compatibility reason we do this only if class does not
+     *   override __new__ ) */
+    __new__ = PyObject_GetAttr((PyObject *)Py_TYPE(self), py___new__);
```

Comment:
@jimfulton thanks for your feedback. You are right. I originally started the patch the way you suggest but it was not working here the way it should - probably some thinko on my side that `__new__` for pythonic objects stored only in type's `tp_dict` and not directly in `tp_new`. I clearly remember I was getting tests failures but now no more; strange... Thanks for catching this.

Updating the patch with the following interdiff:

```diff
diff --git a/persistent/cPersistence.c b/persistent/cPersistence.c
index 3fa2566..67cdbd2 100644
--- a/persistent/cPersistence.c
+++ b/persistent/cPersistence.c
@@ -34,7 +34,6 @@ static PyObject *py___getattr__, *py___setattr__, *py___delattr__;
 static PyObject *py___slotnames__, *copy_reg_slotnames, *__newobj__;
 static PyObject *py___getnewargs__, *py___getstate__;
 static PyObject *py_unsaved, *py_ghost, *py_saved, *py_changed, *py_sticky;
-static PyObject *py___new__, *Persistent__new__;
 
 
 static int
@@ -60,7 +59,6 @@ init_strings(void)
   INIT_STRING(saved);
   INIT_STRING(changed);
   INIT_STRING(sticky);
-  INIT_STRING(__new__);
 #undef INIT_STRING
   return 0;
 }
@@ -153,7 +151,7 @@ accessed(cPersistentObject *self)
 static void
 ghostify(cPersistentObject *self)
 {
-    PyObject **dictptr, *slotnames, *__new__;
+    PyObject **dictptr, *slotnames;
 
     /* are we already a ghost? */
     if (self->state == cPersistent_GHOST_STATE)
@@ -195,9 +193,7 @@ ghostify(cPersistentObject *self)
     /* clear all slots besides _p_*
      * ( for backward-compatibility reason we do this only if class does not
      *   override __new__ ) */
-    __new__ = PyObject_GetAttr((PyObject *)Py_TYPE(self), py___new__);
-    Py_XDECREF(__new__);
-    if (__new__ == Persistent__new__)
+    if (Py_TYPE(self)->tp_new == Pertype.tp_new)
     {
         slotnames = pickle_slotnames(Py_TYPE(self));
         if (slotnames && slotnames != Py_None)
@@ -1510,9 +1506,6 @@ module_init(void)
     Pertype.tp_new = PyType_GenericNew;
     if (PyType_Ready(&Pertype) < 0)
         return NULL;
-    Persistent__new__ = PyDict_GetItem(Pertype.tp_dict, py___new__);
-    if (!Persistent__new__)
-        return NULL;
     if (PyModule_AddObject(module, "Persistent", (PyObject *)&Pertype) < 0)
         return NULL;
```

(with `Py_TYPE` we don't have to ifdef)

Hope the patch is better now. With `git show -w` for C part it is now only:

```diff
diff --git a/persistent/cPersistence.c b/persistent/cPersistence.c
index 5a5bfc2..67cdbd2 100644
--- a/persistent/cPersistence.c
+++ b/persistent/cPersistence.c
@@ -190,7 +190,11 @@ ghostify(cPersistentObject *self)
         *dictptr = NULL;
     }
 
-    /* clear all slots besides _p_* */
+    /* clear all slots besides _p_*
+     * ( for backward-compatibility reason we do this only if class does not
+     *   override __new__ ) */
+    if (Py_TYPE(self)->tp_new == Pertype.tp_new)
+    {
         slotnames = pickle_slotnames(Py_TYPE(self));
         if (slotnames && slotnames != Py_None)
         {
@@ -231,6 +235,7 @@ ghostify(cPersistentObject *self)
             }
         }
         Py_XDECREF(slotnames);
+    }
 
     /* We remove the reference to the just ghosted object that the ring
     * holds.  Note that the dictionary of oids->objects has an uncounted
```

Thanks again,
Kirill